crates/amaru-consensus/src/chain_forward/client_state.rs

```diff
@@ -0,0 +1,119 @@
+pub(super) struct ClientState {
+    store: Arc<Mutex<dyn ChainStore<Header>>>,
+    /// The list of operations to send to the client.
+    ops: VecDeque<ClientOp>,
+    /// The point we presume the client is at.
+    /// This is updated as soon as we send an operation to the client.
+    client_at: Tip,
+}
+
+impl ClientState {
+    pub fn new(
+        store: Arc<Mutex<dyn ChainStore<Header>>>,
+        ops: VecDeque<ClientOp>,
+        client_at: Tip,
+    ) -> Self {
+        Self {
+            store,
+            ops,
+            client_at,
+        }
+    }
+
+    pub async fn next_op(&mut self) -> Option<(ClientOp, Tip)> {
+        let op = self.ops.pop_front()?;
+        let tip = self.tip().await;
+        Some((op, tip))
+    }
+
+    pub async fn tip(&self) -> Tip {
+        if let Some(op) = self.ops.back() {
+            match op {
+                ClientOp::Backward(point) => {
+                    let store = self.store.lock().await;
+                    let header = store.load_header(&hash_point(point)).unwrap();
+                    Tip(point.clone(), header.block_height())
+                }
+                ClientOp::Forward(header) => {
+                    Tip(to_pallas_point(&header.point()), header.block_height())
+                }
+            }
+        } else {
+            self.client_at.clone()
+        }
+    }
+
+    pub fn add_op(&mut self, op: ClientOp) {
+        match op {
+            ClientOp::Backward(_) => todo!(),
+            op @ ClientOp::Forward(_) => self.ops.push_back(op),
+        }
+    }
+}
+
+/// Find headers between points in the chain store.
+/// Returns None if none of the points in `points` lies in the past of `start_point`.
+/// Otherwise returns Some(headers) where headers is a list of headers leading from
+/// the first found point in the past of `start_point` matching a point from `points`
+/// up to `start_point`.
+pub(super) async fn find_headers_between(
+    store: &Arc<Mutex<dyn ChainStore<Header>>>,
+    start_point: &Point,
+    points: &[Point],
+) -> Option<(Vec<ClientOp>, Tip)> {
+    let store = store.lock().await;
+    let start_header = store.load_header(&hash_point(start_point))?;
+
+    // Find the first point that is in the past of start_point
+    let mut current_header = start_header;
+    let mut headers = vec![ClientOp::Forward(current_header.clone())];
+
+    while let Some(parent_hash) = current_header.parent() {
+        match store.load_header(&parent_hash) {
+            Some(header) => {
+                if points.iter().any(|p| hash_point(p) == parent_hash) {
+                    // Found a matching point, return the collected headers
+                    return Some((
+                        headers,
+                        Tip(to_pallas_point(&header.point()), header.block_height()),
+                    ));
+                }
+                headers.push(ClientOp::Forward(header.clone()));
```

Comment:
This is where we'll need to revise the design later to only use indices/pointers to track each followers chain, because otherwise the intersection process could accumulate the whole chain as `Op`s to send to the follower.